### DISCUSSION TITLE
Sort Datatable Columns help

### DISCUSSION THREAD

**Discussion by nastiliano** [flet-dev/flet#3874] [Q&A] [answered]
Hello everybody.

In documentation is not enought explained the right way to sort columns (at least for me).

I understand that I have to edit "sort_column_index" and enable "on_sort event", but...

What kind of code is suposed to put in "on_sort" event?

Is there a little example?

Thank you very much. Best regards.

**Comment by nastiliano** [comment, marked as answer]
Hello everybody,

Attending to this post https://github.com/flet-dev/flet/discussions/1943 I add some lines to the official example in "on_sort" event:



    import flet as ft

    def main(page: ft.Page):
        page.add(
            ft.DataTable(
            width=700,
            bgcolor="yellow",
            border=ft.border.all(2, "red"),
            border_radius=10,
            vertical_lines=ft.BorderSide(3, "blue"),
            horizontal_lines=ft.BorderSide(1, "green"),
            sort_column_index=0,
            sort_ascending=True,
            heading_row_color=ft.colors.BLACK12,
            heading_row_height=100,
            data_row_color={"hovered": "0x30FF0000"},
            show_checkbox_column=True,
            divider_thickness=0,
            column_spacing=200,
            columns=[
                ft.DataColumn(
                    ft.Text("Column 1"),
                    on_sort=lambda e: [# Select the column itself
                                       e.control.parent.__setattr__("sort_column_index" , e.column_index) ,
                                       # Toggle the sort (ascending / descending)
                                       e.control.parent.__setattr__("sort_ascending" , False) if e.control.parent.sort_ascending else e.control.parent.__setattr__("sort_ascending" , True) ,
                                       # Sort the table rows according above
                                       e.control.parent.rows.sort(key=lambda x: x.cells[e.column_index].content.value,reverse = e.control.parent.sort_ascending) ,
                                       # Update table
                                       e.control.parent.update()
                                      ],
                ),
                ft.DataColumn(
                    ft.Text("Column 2"),
                    tooltip="This is a second column",
                    numeric=True,
                    on_sort=lambda e: print(f"{e.column_index}, {e.ascending}"),
                ),
            ],
            rows=[
                ft.DataRow(
                    [ft.DataCell(ft.Text("A")), ft.DataCell(ft.Text("1"))],
                    selected=True,
                    on_select_changed=lambda e: print(f"row select changed: {e.data}"),
                ),
                ft.DataRow([ft.DataCell(ft.Text("B")), ft.DataCell(ft.Text("2"))]),
            ],
        ),
    )

    ft.app(target=main)

Thanks to everybody. Best regards

**Comment by NichtJens** [comment]
The code suggested by @nastiliano brought me on the right track with this problem. I would however suggest some bigger changes:

1. The lambda-returning-a-list "trick" for a multi-line function is usually strongly discouraged.
2. In general, `obj.__setattr__(name, value)` should not be written like this. Instead, use `setattr(obj, name, value)`.
3. The `setattr` in either form should not be used here at all.

So, I would suggest to use this instead:

```python
def on_sort(e):
    p = e.control.parent
    p.sort_column_index = e.column_index
    p.sort_ascending = not p.sort_ascending
    p.rows.sort(key=lambda x: x.cells[e.column_index].content.value, reverse=not p.sort_ascending)
    p.update()
```

This is functionally identical, but syntactically much clearer (I would argue). The only difference is the inserted `not` in `reverse=not ...`. I think the arrow icons in the headers are the wrong way around without this change.

**Comment by ndonkoHenri** [comment]
Fully-fledged example:
```python
import flet as ft


def main(page: ft.Page):
    # Source data for the table (your domain objects). Each record has a stable `id`
    # so we can track selection even when the table is sorted or rebuilt.
    inventory_items = [
        {"id": 1, "name": "Alpha", "qty": 4},
        {"id": 2, "name": "Bravo", "qty": 9},
        {"id": 3, "name": "Charlie", "qty": 2},
        {"id": 4, "name": "Delta", "qty": 6},
        {"id": 5, "name": "Echo", "qty": 3},
        {"id": 6, "name": "Foxtrot", "qty": 8},
        {"id": 7, "name": "Golf", "qty": 1},
        {"id": 8, "name": "Hotel", "qty": 7},
        {"id": 9, "name": "India", "qty": 5},
        {"id": 10, "name": "Juliet", "qty": 10},
    ]

    # Working list used for sorting/reordering. We keep it separate so the original
    # input remains untouched (useful if you later reload or re-filter data).
    displayed_items = list(inventory_items)

    # Store selected item ids (not row indices) so selection survives sorting.
    selected_item_ids: set[int] = {1, 3, 5}

    # Map column index -> callable used for sorting that column.
    # Note: DataColumnSortEvent provides a `column_index` and an `ascending` flag.
    sort_key_for_column = {
        0: lambda item: str(item["name"]).lower(),  # "Item" column
        1: lambda item: int(item["qty"]),  # "Quantity" column
    }

    def build_rows(items: list[dict[str, int | str]]) -> list[ft.DataRow]:
        """Convert a list of item dicts into DataRow objects."""
        return [
            ft.DataRow(
                selected=item["id"] in selected_item_ids,
                on_select_change=handle_row_selection_change,
                data=item["id"],  # used by event handlers to identify this item
                cells=[
                    ft.DataCell(ft.Text(item["name"])),
                    ft.DataCell(ft.Text(str(item["qty"]))),
                ],
            )
            for item in items
        ]

    def refresh_table_rows():
        """
        Rebuild and redraw the table rows.

        Rebuilding rows is the simplest way to keep selection checkboxes and row
        visuals consistent after a bulk change (sort, select-all, clear selection).
        """
        table.rows = build_rows(displayed_items)
        table.update()

    def handle_row_selection_change(e: ft.Event[ft.DataRow]):
        """Called when a single row's checkbox is toggled."""
        row = e.control
        item_id = row.data
        is_selected = e.data  # new selected state

        if is_selected:
            selected_item_ids.add(item_id)
        else:
            selected_item_ids.discard(item_id)

        # Reflect the new state immediately on the toggled row.
        e.control.selected = is_selected
        e.control.update()

    def handle_select_all(e: ft.Event[ft.DataTable]):
        """
        Called when the header "select all" checkbox is toggled.

        `e.data` is True when selecting all, False when clearing.
        """
        select_all = e.data

        if select_all:
            selected_item_ids.update(int(item["id"]) for item in displayed_items)
        else:
            selected_item_ids.clear()

        refresh_table_rows()

    def handle_column_sort(e: ft.DataColumnSortEvent):
        """
        Called when a column header is clicked to sort.

        We sort `displayed_items` in-place and then refresh the rows. Selection is
        preserved because it is tracked by item id in `selected_item_ids`.
        """
        displayed_items.sort(
            key=sort_key_for_column[e.column_index],
            reverse=not e.ascending,
        )

        # Let the Table know which column is currently sorted and in what order.
        table.sort_column_index = e.column_index
        table.sort_ascending = e.ascending

        refresh_table_rows()

    page.add(
        table := ft.DataTable(
            width=700,
            bgcolor=ft.Colors.SURFACE_CONTAINER_LOW,
            border=ft.Border.all(1, ft.Colors.OUTLINE_VARIANT),
            border_radius=10,
            vertical_lines=ft.border.BorderSide(1, ft.Colors.OUTLINE_VARIANT),
            horizontal_lines=ft.border.BorderSide(1, ft.Colors.OUTLINE_VARIANT),
            sort_column_index=0,
            sort_ascending=True,
            heading_row_color=ft.Colors.SURFACE_CONTAINER_HIGHEST,
            heading_row_height=100,
            data_row_color={
                ft.ControlState.HOVERED: ft.Colors.with_opacity(
                    0.08, ft.Colors.PRIMARY
                ),
                ft.ControlState.SELECTED: ft.Colors.with_opacity(
                    0.14, ft.Colors.PRIMARY
                ),
            },
            show_checkbox_column=True,
            on_select_all=handle_select_all,
            divider_thickness=1,
            column_spacing=200,
            columns=[
                ft.DataColumn(
                    label=ft.Text("Item"),
                    on_sort=handle_column_sort,
                ),
                ft.DataColumn(
                    label=ft.Text("Quantity"),
                    tooltip="Numeric quantity",
                    numeric=True,
                    on_sort=handle_column_sort,
                ),
            ],
            rows=build_rows(displayed_items),
        ),
    )


if __name__ == "__main__":
    ft.run(main)
```
